### ISSUE TITLE
Switch add-on verification hashing algorithm to something more appropriate

### ISSUE THREAD

**Issue by maxteufel** [SuperTux/supertux#267]
Change the current MD5 hashes of the add-on ZIPs to SHA-2-512 (even better: SHA-3/Keccak, BLAKE2) since, according to [Wikipedia](https://en.wikipedia.org/wiki/SHA-3#Comparison_of_SHA_functions), MD5 collisions have already been found.
Verifying the hash of archives is important to prevent some attacks using up disk space excessively (zipbombs).

Questions to consider:
- what hashing algorithm (should be SHA-2-512 or better)
- implement it manually as done in `md5.cpp` and `md5.hpp` or switch to a library
- if switching to a library, which one ([mbedTLS](https://tls.mbed.org/), OpenSSL/LibreSSL compatible, GnuTLS)

**Comment by christ2go** [comment]
Honestly I don't believe, that the choice of the hashing algorithm is very important, because it's only used for file integrity checking. The existence of collisions doesn't matter for file integrity checks, only the vulnerability according to preimage attacks is important ([http://security.stackexchange.com/questions/34488/using-md5-for-file-integrity-checks](url)), but it still might be better to switch to a different hash function.
As to which algorithm should be used on one hand I don't see that the files would be so big, that time considerations (MB/s) would matter, on the other hand since it's only used for integrity checking  collisions aren't such a big problem (it wouldn't even be necessary to have a cryptographic hash function e.g. xxHash could be used, it's very good for uniqueness and fast computation).  But any of the functions named above should be perfectly alright, but I'd also consider non-cryptographic hash function, as they are specifically designed for checksum checking.

**Comment by maxteufel** [comment]
@christ2go 

> Verifying the hash of archives is important to prevent some attacks using up disk space excessively (zipbombs).

**Comment by christ2go** [comment]
How would a different hash function offer better zipbomb protection? One would have to be able to create a zip bomb with the same MD5-hash, that one of the addons already has, and then somehow "smuggle" it into github, what also wouldn't be easy. And as I pointed out above it's a differen't thing if collisions exist or if given a file one can easily construct a zipbomb with the same hash.

**Comment by maxteufel** [comment]
Yes one would need to find a zipbomb with a collision. However, the GitHub part is not entirely true since you can also point the game to a different index using the `--repository-url` now (which might not necessarily point to add-ons behind TLS).

**Comment by mstoeckl** [comment]
I've attached a patch which switches to a SHA256 hash function. As all SHA-family hash functions are very similar, changing from SHA-256 to e.g. SHA-512 is not much work. (On one computer, the hash implementation runs at about 120MB/sec, around 1/3 the speed ot md5sum). For this patch to work, SuperTux should be pointed to a version of the `addons` repository that has been rebuilt with the updated `build-addon-index.py` script.

[0002-Switch-addon-verification-hash-to-SHA256.txt](https://github.com/SuperTux/supertux/files/5004154/0002-Switch-addon-verification-hash-to-SHA256.txt)

Let me know if you want me to make a pull request sometime in the next few years :-).
